mgr/dashboard: Improved support for generating OpenAPI Spec documentation

Tinatinak opened the pull request:
<!--
Thank you for opening a pull request!  Here are some tips on creating
a well formatted contribution.

Please give your pull request a title like "[component]: [short description]"

This is the format for commit messages:

"""
[component]: [short description]

[A longer multiline description]

Fixes: [ticket URL on tracker.ceph.com, create one if necessary]
Signed-off-by: [Your Name] <[your email]>
"""

The Signed-off-by line is important, and it is your certification that
your contributions satisfy the Developers Certificate or Origin.  For
more detail, see SubmittingPatches.rst.

The component is the short name of a major daemon or subsystem,
something like "mon", "osd", "mds", "rbd, "rgw", etc. For ceph-mgr modules,
give the component as "mgr/<module name>" rather than a path into pybind.

For more examples, simply use "git log" and look at some historical commits.

This was just a quick overview.  More information for contributors is available here:
https://raw.githubusercontent.com/ceph/ceph/master/SubmittingPatches.rst

-->

- [x] References tracker ticket
- [x] Updates documentation if necessary
- [x] Includes tests for new functionality or reproducer for bug
- [x] Solution finalized
- [x] Linting test pass

Tinatinak commented:
WIP - will fail e.g. linting tests!
Dummy examples in auth.py and cephfs.py are added for demonstrative purpose.

LenzGr commented:
Thanks for submitting! A minor nit: would you mind replacing all occurrences of "respons" with "response", for proper spelling? Thanks :)

rjfd commented on src/pybind/mgr/dashboard/controllers/auth.py:
```diff
@@ -4,20 +4,29 @@
+        param = {
+            'user': {
+                'username': (str, "Description of username"),
+                'password': (str, "Description of password")
```
The reason we followed this approach was because docstrings don't enforce a strict grammar on how to write method documentation. Also, to document a REST API we require more information than just the description of the types of the method parameters and the return type.

These decorators capture what he need to document the REST API without the need to parse docstrings.

sebastian-philipp replied:
It feels as if we're inventing a documentation scheme instead of re-using things that are already there. This isn't a big deal, until we want to automatically extract the docstrings with a different tool like sphinx for the official Ceph documentation. Using this custom style of method documentation will it hard to re-use it. 

Another thing is static type checking as in Python 3 using PEP 484. In addition to duplicating docstrings, we're also duplicating type annotations:
```python
In [1]: def f(x: int): pass

In [3]: f.__annotations__
Out[3]: {'x': int}
```

rjfd replied:
In the future where we only need to support Python 3 we can try to make use of the static type information, but that does not cover all the cases we need to document a REST API. 

For instance, we need to be able to specify a different response JSON structure for each of the HTTP status codes.

sebastian-philipp replied:
Another idea would be to go from the [OpenApi spec to Sphinx](https://sphinxcontrib-openapi.readthedocs.io/). This would solve the problem of having the need to document parameters in docstrings.

Tinatinak commented on src/pybind/mgr/dashboard/controllers/__init__.py:
```diff
@@ -25,6 +25,57 @@
+
+    def _wrapper(func):
+        func.doc_info = {
+            'summary': descr,
```
That's a good idea! But Ḯ'm wondering if this should be added as a 'description' instead, since (I guess?) docstrings may be longer than just one sentence? In the Open API Spec it is possible to describe each endpoint in two ways: 

- 'summary' : the oneliner that should be next to the POST /erasure_code_profile/{name} in the example.
- 'description': a longer description that can be displayed when you open up the endpoint and see more info.  

So far, only 'summary' is implemented. 

![summary and description](https://user-images.githubusercontent.com/22715268/52108979-2dafe800-25fc-11e9-934b-30e604448d11.jpg)
 'description'

Tinatinak commented:
> Thanks for submitting! A minor nit: would you mind replacing all occurrences of "respons" with "response", for proper spelling? Thanks :)

Thanks for pointing that out! I will update!

Tinatinak commented:
Updated PR with a draft for testing. Highly appreciate feedback! 
- Added check for correct format when calling EndpointDoc
- Added new file src/pybind/mgr/dashboard/tests/test_docs.py. Currently it just assure functionality of decorators are OK, but can/should also be used to assure functionality of Docs.py
- I would like to check if swagger page is rendering properly, but I have not found a way to do so!

Tinatinak commented on src/pybind/mgr/dashboard/controllers/auth.py:
```diff
@@ -4,20 +4,29 @@
-
+    @EndpointDoc(
+        descr="DUMMY! Endpoint description", 
+        param = {
```
That is something that came up in one of our meetings, that you guys wanted to be able to define parameters as "dict in dict" (I think I even copied the example from something you wrote! XD). I must have misunderstood? Please explain again!

Tinatinak commented:
Updated PR with suggested changes. Now includes parameters that are arrays. As before: auth.py and cephfs.fy are only included for demonstration and will be removed in final version.

rjfd commented on src/pybind/mgr/dashboard/controllers/docs.py:
```diff
@@ -14,21 +14,56 @@ class Docs(BaseController):
+                    tag_name = ctrl._doc_info['tag']
+                if ctrl._doc_info['tag_descr']:
+                    tag_descr = ctrl._doc_info['tag_descr']
+            tags.append({
```
This will not work when two controllers are using the same "group" but only one of them specifies the "description". You should use a map for the tags variable instead of a list so that in the end you generate a single tag with the correct description.

Tinatinak replied:
Thanks, will change!

Tinatinak commented on src/pybind/mgr/dashboard/controllers/__init__.py:
```diff
@@ -518,7 +634,6 @@ def endpoints(cls):
         """
         This method iterates over all the methods decorated with ``@endpoint``
         and creates an Endpoint object for each one of the methods.
-
```
Hmm... strange... I don't remember removing those lines. I will off course change it back!

Tinatinak commented on src/pybind/mgr/dashboard/controllers/__init__.py:
```diff
@@ -25,6 +25,69 @@
+            resp[str(status_code)] = resp.pop(status_code)
+
+    def _wrapper(func):
+        func.doc_info = {
```
I get a linting error when I have the underscore, so if you don't object I will change it back (without underscore).

rjfd replied:
ok

rjfd commented on src/pybind/mgr/dashboard/controllers/docs.py:
```diff
@@ -99,33 +212,49 @@ def _gen_responses_descriptions(cls, method):
 
+        if resp_object:
+            for status_code, response_body in resp_object.items():
+                resp[status_code].update({
```
you need to cast `status_code` to string `str(status_code)`

Tinatinak replied:
This is already handled in __init__.py and it works for me. Do you get any errors?

rjfd replied:
I had some errors but it was probably my fault. It's working now.

Tinatinak commented:
PR now updated with comments from this morning.

rjfd commented on src/pybind/mgr/dashboard/controllers/docs.py:
```diff
@@ -14,21 +12,61 @@ class Docs(BaseController):
+            if tag_name not in TAG_MAP or not TAG_MAP[tag_name]:
+                TAG_MAP[tag_name] = tag_descr
+
+        tags = []
```
nit: a more pythonic way would be:
```python
tags = [{'name': k, 'description': v if v else "*No description available*"}
        for k, v in TAG_MAP.items()]
```

rjfd commented on src/pybind/mgr/dashboard/tests/test_docs.py:
```diff
@@ -0,0 +1,69 @@
+
+    def test_gen_tags(self):
+        outcome = Docs()._gen_tags(False)[0]
+        self.assertEquals({'description': 'Group description', 'name': 'FooGroup'}, outcome)
```
change `self.assertEquals` to `self.assertEqual` to fix a linting warning.

rjfd commented on src/pybind/mgr/dashboard/tests/test_docs.py:
```diff
@@ -0,0 +1,69 @@
+from .helper import ControllerTestCase
+from ..controllers import RESTController, ApiController, Endpoint, EndpointDoc, ControllerDoc
+from ..controllers.docs import Docs
+
```
Add here the following comment to fix some `protected-access` linting errors:
```python
# pylint: disable=protected-access
```

Tinatinak commented:
Updated PR. Now it includes documentation on how to use the decorators. Suggestions of improvements are always welcome!

Tinatinak commented:
@rjfd For me, code now passes all linting tests (py27-lint and py3-lint). But I wasn't able to reproduce  this linting error. I think the code is OK, but please let me know if you still get this error!
> ```
> ./controllers/__init__.py:69:5: E306 expected 1 blank line before a nested definition, found 0
> ```

callithea commented:
QA run was successful: http://pulpito.ceph.com/laura-2019-02-26_16:37:28-rados:mgr-wip-lpaduano-testing-26227-distro-basic-mira/

LenzGr commented:
Turns out this PR also fixes [bug#38495](https://tracker.ceph.com/issues/38495).
